Replace checkout script with submodules

WalterKolczynski-NOAA opened the pull request:
# Description
Replaces the checkout script with git submodules. This change brings workflow in line with other repos and git standards.

Due to these changes, the build system is updated slightly. First, the options that were formerly used with `checkout.sh` to reduce unneeded builds (`-g` for GSI and `-u` for UFSDA) are moved to `build_all.sh` instead. This is needed since build all will not be able to key off directories existing or not to determine which components to build. Second, the partial build system is removed. It was rarely, if ever, used, and parts of it were already ignored with the previous changes to base builds off whether directories exist. The manage externals file is also removed since that is also not being used.

Following these changes, clones/checkout will need to either be recursive, or `git submodule update` will need to be called separately to make sure components are properly checked out (non-exhaustive list):

```
## Checkout develop (method 1)
git clone --recursive https://github.com/NOAA-EMC/global-workflow destination

## Checkout develop (method 2)
git clone  --jobs 8 https://github.com/NOAA-EMC/global-workflow destination
cd destination
git submodule update --init --recursive --jobs 8

## Checkout existing branch some_branch (method 1)
git clone --recursive  --jobs 8 -b some_branch https://github.com/NOAA-EMC/global-workflow destination

## Checkout existing branch some_branch (method 2)
git clone https://github.com/NOAA-EMC/global-workflow destination
cd destination
git checkout --recurse-submodules some_branch

## Checkout existing branch some_branch (method 3)
git clone https://github.com/NOAA-EMC/global-workflow destination
cd destination
git checkout some_branch
git submodule update --init --recursive  --jobs 8

## Checkout new branch new_branch
git clone https://github.com/NOAA-EMC/global-workflow destination
cd destination
git checkout -b new_branch
git submodule update --init --recursive  --jobs 8
```

The git ignore list had to be updated to use explicit paths as the wildcard paths pick up the ones that are now submodules.

UFS creates files during the build process that are not in its ignore list, so that submodule ignores the "dirty" state to avoid confusing users running `git status`.

CI and documentation are also updated to reflect the changes.

Resolves #2141 

# Type of change
- Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? YES
- Does this change require a documentation update? YES

# How has this been tested?
- Clone and build on Orion

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] I have made corresponding changes to the documentation if necessary

WalterKolczynski-NOAA committed:
Add submodules to replace checkout script

Git submodules are added with the intention of replacing the checkout.sh
script. Some changes (like deleting the checkout script) are not yet
completed pending testing.

WalterKolczynski-NOAA committed:
Delete checkout and update git ignore

Removes the checkout script that has been replaced by git submodules

Finishes the updates to the .gitignore since we don't want wildcards to
cover the new submodule directories.

WalterKolczynski-NOAA committed:
Remove manage externals and partial build

Removes the manage externals config since it is no longer an option
after switching to submodules.

Also removes the script that parses the "partial build" config and
the config. This feature was unused and had already been partially
deactivated to look for directory existance instead.

WalterKolczynski-NOAA committed:
Add build options to control build GSI and UFSDA

The move to submodules eliminates the capability to build based on
the existance of directories, since they will now always be checked
out. To avoid unnecessary compilation time, the build script is
updated to take the same options as the old checkout script to build
GSI (-g) and UFSDA (-u). The default is to not build these.

WalterKolczynski-NOAA committed:
Update documentation for submodules

WalterKolczynski-NOAA committed:
More documentation updates for submodule change

WalterKolczynski-NOAA committed:
Update CI for submodule change

WalterKolczynski-NOAA committed:
Remove one commented-out git ignore

WalterKolczynski-NOAA committed:
Ignore dirty ufs_model submodule

UFS creates some files that are not in the repo's ignore list. This
change will ignore those and avoid confusing users running `git status`.

WalterKolczynski-NOAA committed:
Slight adjustment to CI checkout

WalterKolczynski-NOAA commented:
@TerrenceMcGuinness-NOAA we'll need you to test the new CI works properly. With this type of change, I don't think there is much use in sending this through the normal CI channel; they'll have to be updated simultaneously.

KateFriedman-NOAA commented on docs/source/clone.rst:
```diff
@@ -9,41 +9,40 @@ Quick Instructions
 
 .. note::
-   Here we are making the assumption that you are using the workflow to run an experiment and so are working from the authoritative repository. If you are using a development branch then follow the instructions in :doc:`development.rst`. Once you do that you can follow the instructions here with the only difference being the repository/fork you are cloning from.
+   Here we are making the assumption that you are using the workflow to run an experiment and so are working from the authoritative repository . If you are using a development branch then follow the instructions in :doc:`development.rst`. Once you do that you can follow the instructions here with the only difference being the repository/fork you are cloning from.
```
Remove the errant space that snuck in.

WalterKolczynski-NOAA replied:
Fixed

KateFriedman-NOAA reviewed (approved):
Looks great, thanks for working on this @WalterKolczynski-NOAA ! I left one tiny comment to remove an errant space. No other comments.

DavidHuber-NOAA reviewed (changes requested):
Looks pretty good.  I ran a checkout/build on Orion and everything worked as expected.  The only thing I see is that the link script will create dead links to anything that isn't built.  Perhaps the check for GDAS should be `[[ -d "${HOMEgfs}/sorc/gdas.cd/bin" ]]` and for the GSI, GSI-monitor, and GSI-utils: `[[ -d "${HOMEgfs}/sorc/<gsi_dir>/install" ]]`.  Otherwise, this looks really great!

WalterKolczynski-NOAA committed:
Remove errant space in documentation

HenryRWinterbottom commented:
LGTM. Thanks @WalterKolczynski-NOAA.

aerorahul commented:
I don't think it is this PR, but there are some dead links referring to GDASApp executables after `link_workflow.sh` is executed.  See the `red` ones in the image below:

![Screenshot 2023-12-12 at 12 33 13 PM](https://github.com/NOAA-EMC/global-workflow/assets/11394126/b9eacf18-0dcf-45a7-b33b-9f2bbef6c859)

@CoryMartin-NOAA @RussTreadon-NOAA Are these being built correctly in the GDASApp?  Are they being used in the global applications?

aerorahul reviewed (approved):
Looks good.
I don't see the issues @DavidHuber-NOAA reported.

CoryMartin-NOAA commented:
yeah I think names changed, @aerorahul, we can fix it "soon". My priority this afternoon/tomorrow is to get GDASApp to also use submodules. I will fix this in the PR that adds the new submodule for GDASApp.

aerorahul commented:
> yeah I think names changed, @aerorahul, we can fix it "soon". My priority this afternoon/tomorrow is to get GDASApp to also use submodules. I will fix this in the PR that adds the new submodule for GDASApp.

no rush.  Just noting something has changed.

DavidHuber-NOAA commented:
@aerorahul I cloned the branch, built just the core programs (i.e. no DA), and then linked the workflow as follows:

```
git clone git@github.com:WalterKolczynski-NOAA/global-workflow -b feature/submodules --recursive
cd global-workflow/sorc
./build_all.sh
./link_workflow.sh
```

This resulted in several dead links in the exec directory as shown below.  The `link_workflow.sh` script only looks for the presence of the head of the directories (e.g. `gdas.cd`) and links based on that rather than what was built.  Since only a subset of users will build the GDASApp, I would prefer that those links only be created if a user is planning on using that system.  Similarly, some users may not want to build DA at all and only run forecasts or build the GDASApp without the GSI.
![Screenshot 2023-12-12 135046](https://github.com/NOAA-EMC/global-workflow/assets/69919478/719656fd-33e5-4504-84e0-c08142e1c59a)

WalterKolczynski-NOAA commented:
> @aerorahul I cloned the branch, built just the core programs (i.e. no DA), and then linked the workflow as follows:
> 
> ```
> git clone git@github.com:WalterKolczynski-NOAA/global-workflow -b feature/submodules --recursive
> cd global-workflow/sorc
> ./build_all.sh
> ./link_workflow.sh
> ```
> 
> This resulted in several dead links in the exec directory as shown below. The `link_workflow.sh` script only looks for the presence of the head of the directories (e.g. `gdas.cd`) and links based on that rather than what was built. Since only a subset of users will build the GDASApp, I would prefer that those links only be created if a user is planning on using that system. Similarly, some users may not want to build DA at all and only run forecasts or build the GDASApp without the GSI. ![Screenshot 2023-12-12 135046](https://private-user-images.githubusercontent.com/69919478/289972524-719656fd-33e5-4504-84e0-c08142e1c59a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI0MTI0OTksIm5iZiI6MTcwMjQxMjE5OSwicGF0aCI6Ii82OTkxOTQ3OC8yODk5NzI1MjQtNzE5NjU2ZmQtMzNlNS00NTA0LTg0ZTAtYzA4MTQyZTFjNTlhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjEyVDIwMTYzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY1OWE1ZWJmOTc0NWMxNDQzNTE1YTgyNzUxZmVhZjc4N2ZjMTFhOWJmZDNkNTgxNGY1OTIzM2M0OTZjODZjZjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.QSSHAI3ND0yMSmzc0jb8ZQTyy_jAIKRLCa_k5sTRBqk)

Yeah, I'm planning on addressing this, I have just been stuck in meetings all day.

WalterKolczynski-NOAA committed:
Stop making links for unbuilt components

After switching to submodules, the link script could no longer use
the existence of directories to determine which executables to link.
Now it instead looks for the existence of the build directory.

Refs #2141

WalterKolczynski-NOAA commented:
@DavidHuber-NOAA Broken links when skipping UFSDA/GSI build should be gone now.

DavidHuber-NOAA reviewed (approved):
Looks good, thanks @WalterKolczynski-NOAA!

DavidHuber-NOAA commented:
This is beyond the scope of this PR, but I noted that there are some other broken links in the workflow, noted below

![image](https://github.com/NOAA-EMC/global-workflow/assets/69919478/16615373-aad2-4f3f-92b4-dc2bda9a11c8)

DavidHuber-NOAA commented:
I looked into the CI test failure.  It was another timeout of `radmon_time.x`.  I noted this one other time.  Perhaps there is an intermittent hang?

aerorahul commented:
> This is beyond the scope of this PR, but I noted that there are some other broken links in the workflow, noted below
> 
> ![image](https://private-user-images.githubusercontent.com/69919478/290205050-16615373-aad2-4f3f-92b4-dc2bda9a11c8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI0ODIzNDcsIm5iZiI6MTcwMjQ4MjA0NywicGF0aCI6Ii82OTkxOTQ3OC8yOTAyMDUwNTAtMTY2MTUzNzMtYWFkMi00ZjNmLTkyYjQtZGMyYmRhOWExMWM4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjEzVDE1NDA0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFiNGQ3YWQzNzI2NGYwMTQ4NzE1YmQ0Y2UyY2VlMDc1YTkyNGMzZTMwM2YzNTU5MWI4YzY0NTBlZDZlOGRiYjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.-rwm6aAonStTVzcFuH-pwRVnRd9xIsJfEA7VOTr2R2U)

The `global_angupdate` dead link is because of someone named @aerorahul might have tried to not duplicate them.  Since the packages of gsi and gsi-utils are separated, it might need duplication.  On the other hand, global_angupdate has been long deprecated and remains a relic in gsi-utils repository.

TerrenceMcGuinness-NOAA commented:
```
mterry (Orion-login-1) global-workflow (feature/submodule) $ clear
mterry (Orion-login-1) global-workflow (feature/submodule) $ source versions/build.spack.ver
mterry (Orion-login-1) global-workflow (feature/submodule) $ source versions/run.orion.ver
mterry (Orion-login-1) global-workflow (feature/submodule) $ module load module_base.orion
mterry (Orion-login-1) global-workflow (feature/submodule) $ ldd exec/radmon_time.x | head
	linux-vdso.so.1 =>  (0x00007ffccff8f000)
	libnetcdff.so.7 => /work/noaa/epic/role-epic/spack-stack/orion/spack-stack-1.5.1/envs/unified-env/install/intel/2022.0.2/netcdf-fortran-4.6.0-i2wplrg/lib/libnetcdff.so.7 (0x00007fc2ddc79000)
	libnetcdf.so.19 => /work/noaa/epic/role-epic/spack-stack/orion/spack-stack-1.5.1/envs/unified-env/install/intel/2022.0.2/netcdf-c-4.9.2-j2wl3gb/lib/libnetcdf.so.19 (0x00007fc2dd7fb000)
	libm.so.6 => /lib64/libm.so.6 (0x00007fc2dd4f9000)
	libmpifort.so.12 => /apps/intel-2022.1.2/intel-2022.1.2/mpi/2021.5.1/lib/libmpifort.so.12 (0x00007fc2dd145000)
	libmpi.so.12 => /apps/intel-2022.1.2/intel-2022.1.2/mpi/2021.5.1/lib/release/libmpi.so.12 (0x00007fc2db910000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007fc2db70c000)
	librt.so.1 => /lib64/librt.so.1 (0x00007fc2db504000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fc2db2e8000)
	libc.so.6 => /lib64/libc.so.6 (0x00007fc2daf1b000)
mterry (Orion-login-1) global-workflow (feature/submodule) $ time exec/radmon_time.x --help
```

Looks line `radmon_time.x` is hanging --- no, I see it is waiting for input, never-mind. Waiting for other's input.

WalterKolczynski-NOAA committed:
Extend verifrad wall clock

Verfrad is hitting the wall clock during CI testing, so it is doubled
temporarily while the root cause is discovered.

WalterKolczynski-NOAA commented:
Orion is clearly unstable again, as it is failing on a different job each time.

WalterKolczynski-NOAA commented:
Remaining experiment is only waiting for archive and cleanup, but is stuck queued because of the imminent maintenance downtime. Going to set this to passed manually and merge.
